cmd/collectors/ems/ems_test.go

```diff
@@ -0,0 +1,241 @@
+			if !ok && generatedEmsName != "Ems" {
+				t.Fatalf("Failed to get netric for Ems %s", generatedEmsName)
+			}
+			for _, instance := range mx.GetInstances() {
+				// If value not exist for that instance or metric value 0 indicate ems hasn't been raised.
+				if val, ok := metr.GetValueFloat64(instance); !ok || val == 0 {
+					notGeneratedEmsNames = append(notGeneratedEmsNames, generatedEmsName)
+				}
+				// Test for matches - filter
+				if generatedEmsName == "hm.alert.raised" {
+					if instance.GetLabel("alert_id") == "RaidLeftBehindAggrAlert" {
+						// OK
+					} else {
+						t.Errorf("Labels alert_id= %s, expected: RaidLeftBehindAggrAlert", instance.GetLabel("alert_id"))
+					}
+				}
+			}
+		} else if generatedEmsName != "Ems" {
+			t.Errorf("Extra bookend issuing ems event found= %s", generatedEmsName)
+		}
+	}
+	count := len(notGeneratedEmsNames)
+	if count > 0 {
+		t.Fatalf("These Bookend Ems haven't been raised: %s", notGeneratedEmsNames)
+	}
+
+	// Simulated bookend resolving ems "wafl.vvol.online" and ems "hm.alert.cleared" with alert_id value as "RaidLeftBehindAggrAlert"
+	results = jsonToEmsRecords(t, "testdata/ems-poll-3.json")
+	// Polling ems collector to handle results
+	emsMetric, emsCount = e.HandleResults(results, e.emsProp)
+	if emsCount == 0 {
+		t.Fatal("Failed to fetch data")
+	}
+
+	// Check and evaluate ems events
+	var notResolvedEmsNames []string
+	for generatedEmsName, mx := range emsMetric {
+		if util.Contains(issuingEmsNames, generatedEmsName) {
+			metr, ok := mx.GetMetrics()["events"]
+			// e.Matrix map would have one entry of Ems(parent) in map, skipping that as it's not required for testing.
+			if !ok && generatedEmsName != "Ems" {
+				t.Fatalf("Failed to get netric for Ems %s", generatedEmsName)
+			}
+			for _, instance := range mx.GetInstances() {
+				// If value exist for that instance and metric value 1 indicate ems hasn't been resolved.
+				val, ok := metr.GetValueFloat64(instance)
+				if ok && val == 1 {
+					notResolvedEmsNames = append(notResolvedEmsNames, generatedEmsName)
+				}
+				// Test for matches - filter
+				if generatedEmsName == "hm.alert.raised" {
+					if instance.GetLabel("alert_id") == "RaidLeftBehindAggrAlert" && ok && val == 0.0 {
+						// OK
+					} else {
+						t.Errorf("Labels alert_id= %s, expected: RaidLeftBehindAggrAlert, metric value = %f, expected: 0.0", instance.GetLabel("alert_id"), val)
+					}
+				}
+			}
+		} else if generatedEmsName != "Ems" {
+			t.Errorf("Extra bookend issuing ems event found= %s", generatedEmsName)
+		}
+	}
+	// After resolving ems, all bookend ems should resolved
+	count = len(notResolvedEmsNames)
+	if count > 0 {
+		t.Errorf("These Bookend Ems haven't been resolved: %s", notResolvedEmsNames)
+	}
+}
+
+func Test_Ems(t *testing.T) {
+	e := Setup()
+	NonBookendEmsTest(t, e)
+	BookendEmsTest(t, e)
+}
+
+func newEms(homePath string) *Ems {
```

Comment:
minor but I don't think the poll or auth are needed. This should work

```go
func newEms(homePath string) *Ems {
	conf.TestLoadHarvestConfig("testdata/config.yml")
	opts := options.Options{
		Poller:   EmsPollerName,
		HomePath: homePath,
	}
	ac := collector.New("Ems", "Ems", &opts, emsParams(), nil)
	ac.IsTest = true
	e := &Ems{}
	err := e.Init(ac)
	if err != nil {
		log.Fatal().Err(err)
	}
	return e
}

```

Reply:
Done